src/xcode/ENA/ENA/Source/Services/HealthCertificateValidationService/HealthCertificateValidationService.swift

```diff
@@ -172,4 +212,557 @@ final class HealthCertificateValidationService: HealthCertificateValidationProvi
+				completion(.success(cachedAcceptanceRules))
+			} else {
+				// If not, return edge case error
+				Log.error("Could not find cached acceptance rules but need some.", log: .vaccination)
+				completion(.failure(.ACCEPTANCE_RULE_MISSING_CACHE))
+			}
+		case .noNetworkConnection:
+			Log.error("Could not download acceptance rules due to no network.", log: .vaccination, error: error)
+			completion(.failure(.NO_NETWORK))
+		case let .serverError(statusCode):
+			switch statusCode {
+			case 400...409:
+				Log.error("Could not download acceptance rules due to client error.", log: .vaccination, error: error)
+				completion(.failure(.ACCEPTANCE_RULE_CLIENT_ERROR))
+			default:
+				Log.error("Could not download acceptance rules due to server error.", log: .vaccination, error: error)
+				completion(.failure(.ACCEPTANCE_RULE_SERVER_ERROR))
+			}
+		default:
+			Log.error("Could not download acceptance rules due to server error.", log: .vaccination, error: error)
+			completion(.failure(.ACCEPTANCE_RULE_SERVER_ERROR))
+		}
+	}
+	
+	private func acceptanceRules(_ data: Data, completion: (Result<[Rule], HealthCertificateValidationError>) -> Void) {
+		let extractAcceptanceRulesResult = ValidationRulesAccess().extractValidationRules(from: data)
+		
+		switch extractAcceptanceRulesResult {
+		case let .success(acceptanceRules):
+			completion(.success(acceptanceRules))
+		case .failure:
+			completion(.failure(.ACCEPTANCE_RULE_JSON_DECODING_FAILED))
+		}
+	}
+	
+	// MARK: - Invalidation Rules
+	
+	private func downloadInvalidationRule(
+		completion: @escaping (Result<[Rule], HealthCertificateValidationError>) -> Void
+	) {
+		client.getDCCRules(
+			eTag: store.invalidationRulesCache?.lastValidationRulesETag,
+			isFake: false,
+			ruleType: .invalidation,
+			completion: { [weak self] result in
+				guard let self = self else {
+					Log.error("Could not create strong self", log: .vaccination)
+					completion(.failure(.INVALIDATION_RULE_CLIENT_ERROR))
+					return
+				}
+				
+				switch result {
+				case let .success(packageDownloadResponse):
+					self.invalidationRuleDownloadingSuccessHandler(
+						packageDownloadResponse: packageDownloadResponse,
+						completion: completion
+					)
+	
+				case let .failure(error):
+					self.invalidationRuleDownloadingFailureHandler(
+						error: error,
+						completion: completion
+					)
+				}
+			}
+		)
+	}
+	
+	private func invalidationRuleDownloadingSuccessHandler(
+		packageDownloadResponse: PackageDownloadResponse,
+		completion: @escaping (Result<[Rule], HealthCertificateValidationError>) -> Void
+	) {
+		Log.info("Successfully received invalidation rules package. Proceed with eTag verification...", log: .vaccination)
+
+		guard let eTag = packageDownloadResponse.etag else {
+			Log.error("ETag of package is missing. Return with failure.", log: .vaccination)
+			completion(.failure(.INVALIDATION_RULE_JSON_ARCHIVE_SIGNATURE_INVALID))
```

Comment:
Same question here, if this is the correct error code

Reply:
Same response: new error case for this case.